### PR TITLE
Improve ztest macros

### PR DESCRIPTION
**Pull request by barnas-michal**
These changes removes the requirement for sending the NULL as last parameter if the message is not used in zassert_* and zassume_* macros.

If the default message is descriptive enough, there's no need to provide the additional one, but it was required to send NULL as parameter to the macros. It will no longer be needed after this change.

### CODE REVIEW

**Review by yperess — CHANGES_REQUESTED** [review]
Can you update some tests and also include a sample error message if none was specified?

**Review by jeremybettis — DISMISSED** [review]

File: subsys/testsuite/ztest/include/zephyr/ztest_assert.h
Comment (aaronemassey) [inline subsys/testsuite/ztest/include/zephyr/ztest_assert.h:159]: This will conflict with https://github.com/zephyrproject-rtos/zephyr/pull/48667/files which I think will be merged today. So may want to hold off and rebase.

**Review by aaronemassey — DISMISSED** [review]
LGTM, but should wait to rebase off those doc changes I'm about to merge I think.

File: subsys/testsuite/ztest/include/zephyr/ztest_assert.h
Comment (semihalf-barnas-michal) [reply subsys/testsuite/ztest/include/zephyr/ztest_assert.h:159]: Good to know, I will wait

**Comment by semihalf-barnas-michal** [comment]
> Can you update some tests and also include a sample error message if none was specified?
Sure, I've added it also to the commit message of the one that updates the zassert_* macros.

```
void* beauty_ptr = NULL;
zassert_not_null(beauty_ptr);
```
=>
`Assertion failed at main.c:20: suite_test: beauty_ptr is NULL`

```
void* ugly_ptr = (void*)0xbaadf00d;
zassume_is_null(ugly_ptr);
```
=>
`Assumption failed at main.c:23: suite_test: ugly_ptr is not NULL`

```
int apple = 3;
int tomato = 9;
zassume_equal(apple, tomato);
```
=>
`Assumption failed at main.c:27: suite_test: apple not equal to tomato`

File: tests/unit/cbprintf/main.c
Comment (stephanosio) [inline tests/unit/cbprintf/main.c:587]: Please leave `%Ld` here as is. From what I can see, this is intentionally testing an invalid format specifier.

**Review by stephanosio — DISMISSED** [review]

File: tests/unit/cbprintf/main.c
Comment (semihalf-barnas-michal) [reply tests/unit/cbprintf/main.c:587]: Yes, you're right. I didn't revert this patch yet, because I am looking for a help with suppressing the checkpatch warning about this format.

**Review by yperess — CHANGES_REQUESTED** [review]
It would be nice to have actual framework tests for this. I think you could add a test under `tests/ztest/assert` that do something like:
```
ZTEST_EXPECT_FAIL(asserts, test_default_assert_failure_message)
ZTEST(asserts, test_default_assert_failure_message)
{
  zassert_false(true);
}
```

Then in that test's testcase.yaml you can use the harness_config like:
```
harness_config:
  type: multi_line
  ordered: false
  regex:
    - "<the default error message>"
```

(I believe that this should all work)

**Review by yperess — DISMISSED** [review]

**Review by yperess — DISMISSED** [review]
Spoke with Michael offline and he's actively working on being able to use the console test harness with ztest. Having that will allow us to create tests that assert on the error message. Since these tests never existed, I don't see a reason to block this PR as it makes the assert and assume API much better.

**Comment by semihalf-barnas-michal** [comment]
> It would be nice to have actual framework tests for this
I will do them in separate PR, since it requires to fix the regexes verification in the ztest framework.

This PR has compliance check failing due to the test that checks for handling of invalid printf format string. This is a false-positive check and should be ignored.
`WARNING:PRINTF_L: %Ld is non-standard C, use %lld`

**Review by andyross — DISMISSED** [review]
Not sure if I ever reviewed this one or not, but I love the feature.  Care to play the same trick with __ASSERT() so we can eliminate ___ASSERT_NO_MSG()? :)

**Comment by semihalf-barnas-michal** [comment]
> Not sure if I ever reviewed this one or not, but I love the feature. Care to play the same trick with __ASSERT() so we can eliminate ___ASSERT_NO_MSG()? :)

Sure, I can take a look at this later :)

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
CI fail is unrelated:

````
#15420: FILE: tests/unit/cbprintf/main.c:527:
+	zassert_equal(strncmp("/%Ld/", buf, rc), 0);
```

That'll have to be bypassed.

File: subsys/testsuite/ztest/include/zephyr/ztest_assert.h
Comment (henrikbrixandersen) [inline subsys/testsuite/ztest/include/zephyr/ztest_assert.h:150]: Is `__VA_OPT__` supported by all the toolchains supported by Zephyr? /cc @stephanosio

**Review by henrikbrixandersen — DISMISSED** [review]

File: subsys/testsuite/ztest/include/zephyr/ztest_assert.h
Comment (stephanosio) [reply subsys/testsuite/ztest/include/zephyr/ztest_assert.h:150]: For one, `__VA_OPT__` is non-standard and a GNU extension.

I am not sure if it is supported by the XCC, which is based on GCC 4.x. Can you confirm? @nashif @dcpleung @laurenmurphyx64

Comment (nashif) [reply subsys/testsuite/ztest/include/zephyr/ztest_assert.h:150]: looks like it is.

**Review by stephanosio — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@stephanosio can you manually merge this?

**Comment by stephanosio** [comment]
> @stephanosio can you manually merge this?

uhh .. looks like we have a conflict now.

**Comment by stephanosio** [comment]
@semihalf-barnas-michal can you please rebase?

**Review by fabiobaltieri — APPROVED** [review]
@stephanosio this should be good to go now.

**Comment by semihalf-barnas-michal** [comment]
> can you please rebase?

@stephanosio : Sure, done. The only warning is a false-positive against the sprintf format test.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by nashif** [comment]
> @stephanosio this should be good to go now.

it is only good to go when reviewed by assignee, whoever that might be :-)

**Review by nashif — APPROVED** [review]

**Pull request merged**
